tracer/src/Datadog.Trace/Propagators/W3CTraceContextPropagator.cs

```diff
@@ -6,129 +6,565 @@
-                    base16(trace-id) = 4bf92f3577b34da6a3ce929d0e0e4736
-                    base16(parent-id) = 00f067aa0ba902b7
-                    base16(trace-flags) = 00  // not sampled
-                 */
+                            sb.Append("t.").Append(key).Append(':').Append(tag.Value).Append(';');
+                        }
+                    }
+                }
 
-                if (traceParent!.Length != 55 || traceParent[2] != '-' || traceParent[35] != '-' || traceParent[52] != '-')
+                if (sb.Length == 3)
                 {
-                    return false;
+                    // "dd=", we never appended anything
+                    return string.Empty;
                 }
 
-                char w3cSampled = traceParent[54];
-                if (traceParent[53] != '0' || (w3cSampled != '0' && w3cSampled != '1'))
+                // remove trailing ";"
+                if (sb[sb.Length - 1] == ';')
                 {
-                    return false;
+                    sb.Length--;
                 }
 
-                var samplingPriority = w3cSampled == '0' ? 0 : 1;
+                return sb.ToString();
+            }
+            finally
+            {
+                StringBuilderCache.Release(sb);
+            }
+        }
+
+        internal static bool TryParseTraceParent(string header, out W3CTraceParent traceParent)
+        {
+            // "{version:2}-{trace-id:32}-{parent-id:16}-{trace-flags:2}"
+            //             ^ 2           ^ 35           ^ 52            ^ 55
+
+            traceParent = default;
+
+            if (header == null!)
+            {
+                return false;
+            }
+
+            header = header.Trim();
+
+            if (header.Length < 55 || header[2] != '-' || header[35] != '-' || header[52] != '-')
+            {
+                // too short, or invalid delimiter positions
+                return false;
+            }
+
+            if (header[0] < '0' || header[0] > 'f' || header[1] < '0' || header[1] > 'f')
+            {
+                // invalid version value, must contain lower-case hexadecimal characters
+                return false;
+            }
+
+            if (header[0] == 'f' && header[1] == 'f')
+            {
+                // while "ff" is valid hex, it is explicitly not allowed as a version value
+                return false;
+            }
+
+            if (header[0] == '0' && header[1] == '0' && header.Length != 55)
+            {
+                // for version "00", the length must be exactly 55
+                return false;
+            }
+
+            if (header.Length > 55 && header[55] != '-')
+            {
+                // if there is more data than expected (e.g. future version of the spec),
+                // it's should to be additive, so there must be another delimiter after `trace-tags`
+                return false;
+            }
+
+            ulong traceId;
```

Comment:
Shall we do all that only in the case of version 00 to be future proof? The RFC says:

> If the version is “00”, parse and validate the [trace-id](https://docs.google.com/document/d/1xacBonCyuVk95D-L1STXGqdLtsOuf0_3jG-138uXHnQ/edit#heading=h.bhz2cebri5hn), the [parent-id](https://docs.google.com/document/d/1xacBonCyuVk95D-L1STXGqdLtsOuf0_3jG-138uXHnQ/edit#heading=h.30k09gwvr1sx), and the [sampled flag](https://docs.google.com/document/d/1xacBonCyuVk95D-L1STXGqdLtsOuf0_3jG-138uXHnQ/edit#heading=h.tyfsywks04ci) in the trace-flags (right-most bit). If any are invalid, ignore the tracestate header and move on to the next propagator.

Reply:
Not sure what you meant since this comment is on line `ulong traceId;`, but you quoted section "Header extraction logic > 3 > b". Section `d` below explains what to do if the version is not `00` but is a valid future version (not `ff`):
> Otherwise, if the version can be parsed as two hex digits, follow these parsing rules...

Reply:
The reason is that the W3C expected versions to be additive, so a version `01` for example would include everything in version `00` plus additional data at the end after position 55.

Reply:
> Not sure what you meant since this comment is on line

Yes I meant all the following code, sorry.

> Otherwise, if the version can be parsed as two hex digits, follow these parsing rules

Ok I had missed that. Thanks